### ISSUE TITLE
Main castle defenders and swapping knights

### ISSUE THREAD

**Issue by Wakeboardtiger** [nicymike/serflings#20]
Seems like the logic of exchanging defenders in the main castle is not working correctly. I am having a hard time to get the good guys out of school and the new guys into it, by pressing the knight exchange button on the recruit page. Many guys are coming and in the end the good guys stayed behind. I tried to relive em from the defender page, as well so they staid in castle ready to leave, but same result. I will investigate further.
And also a small detail: the percentage sign is missing on the recruiting page above the available gold bar number.

But let me thank you Nicymike and everyone who is been working on this. I am dreaming of a realization of this game for multiplayer's on network my lifelong. This is an amazing project and has been greatly progressed in this direction. I would love to help, but i have hardly any knowledge of programming, unfortunately.  (But maybe never to late, let me know)
But so, may I address a wish here, to maybe focus on the full functionality on the human players first, before moving on to the AI.
From what I gather there are still known missing tools from the original game like on the website mentioned:

Original game: Connect a flag to an already existing path
Original game: Bird sounds, 
Original game: Emergency program with too little building materials at the start of the game
Original game: Win or loose a game
let me add: 
Original game: change building type by clicking on existing building and choose on the page the new building, confirm with both button pressed, to tear down the old and create the building stone for the new planned one.

which are not realized yet, but which are for a player who knows the game are essential, to enjoy the game. You surely have your reasons to delay these, due to the complexity of the problems here, i am not able to comprehend :o) But just to have it said, that we are in need of those, maybe before we start playing against the proper AI. Bcs in general, if i want to play against the computer guys, i can still use the original game and DosBox in SVGA which is very well playable still, without any surprises.
But with serflings i might be able to compete against or play in one team ( if you implement ) with friends, on the network or in the future on the internet against real people, which makes the whole difference to the original game. Therefore, please also keep it as original as practical to see how well we master this original game in comparison to others. ;o) same chances for everyone who played the original game before. 
Of course a view bug fixes from the old game are also welcome, like sometimes with multiple warehouses some warehouses get stuck after time and keep everything inside, not even being able to force them to move all the goods out, which i just experienced in serflings as well. ( I can sent you a copy of a safed game of mine if it helps )
So please keep up the great work, and let me know how i can help the project.
Many thanks

**Comment by nicymike** [comment]
Thank you for the big feedback. I will try to answer to all of your subjects separately. If I forget something just give me a comment.

1. Whenever you exchange knights the new knights will most likely come from the nearest stock even if there are better knights in a different stock. At least this is what I got from the original source.

2. About the missing percentage sign: Can you attach a screenshot because the sign should be there.

3. The best way to help me is to test and play the game and to spread the word about it. I appreciate every feedback because I want to make a game which is fun to play even today.

4. About the AI: I wrote with some people on youtube about the game and I often got the answer: "Nice game but I want to wait until AI is finished (for playing missions)." So I decided to focus on the AI first which is btw. almost finished because only attacking is left.
My plan is to get the game ready to play missions till the end of 2020. This also includes some of the other stuff missing from the original game. I like to see Serflings beeing played on youtube let's plays.

5. Let me go through the open tasks from the original game you mentioned:
- Connect a flag to an existing path:
Actually the logic behind this is already implemented because the AI needed this. :) Only the action for the player is missing.
- Bird sounds:
I tried to add this but the sounds went crazy. I need to add the priority system for sounds of the original game to reduce the amount of sounds playing at the same time. Think of the burn sound in 8x speed (or the sawmill).
- Emergency program:
The main issue is that there is logic for this spread all over the code. The stocks need to work with this to priorise goods, the construction sites need to take it into account for requesting workers and so on. I know this is essential but I think I will need more time to add this.
- Replace an existing building:
Oh, I missed this one on the website. Thank you for mention it. I did not look into this yet.

6. I did not know about stocks getting stuck. Please attach a savegame. I guess I need to check the code.

7. The original game is ok but it is also really choppy. Just think of mouse movement and scrolling. Even the history edition did not improve this (and lost the transparency on shadows instead).
There are not many people out there playing Serflings multiplayer (at least I don't know many). There are some multiplayer games on youtube that I am really happy about. But multiplayer will surely be one of the main parts of the game. I will work on this after finishing the AI and playable missions. I have plans about optimizing the network code to allow up to four players. And I will also try to keep the game as original as possible and plan to add a setting for every difference I add.

It is difficult to meet all the people's expectations, so I have to decide on which issue I want to work next. This is even harder since I am only a single person working on this in my spare time. :) But I like settlers 1 and I like working on Serflings.

Thank you again for the great feedback

**Comment by Wakeboardtiger** [comment]
Thanks for the detailed answer. I feel bad keeping you busy with that. :)
So here for Nr.2 the screenshot; I think the problem appears above 100%, three digits.
![image](https://user-images.githubusercontent.com/68932712/88953622-639e5d00-d2c3-11ea-8a73-a0e782683e88.png)

left original, right serflings.


To Nr.1:
![image](https://user-images.githubusercontent.com/68932712/88951368-153b8f00-d2c0-11ea-85ad-47563234185c.png)

I ran a quick test again and the result, as you can see the big knights left on the original game and where still in school in serflings after the swap. So in the original they walk into the castle as freshmen and walk out as proper kinghts, while in serflings they walk in the same they walk out. There are still some strange things going on, which dont match in the school. :)
- Another difference is, when I want to open school places, in serflings, knights are placed instantly into school while in the original it takes a bit a time to get em there. Same effect when you dismiss them from school.
![image](https://user-images.githubusercontent.com/68932712/88953158-b0cdff00-d2c2-11ea-83d8-f6502aede09f.png)
- And above the diffrence when you add em back to school, original puts the freshmen first serflings is taking the proper knights back into school.
And more small details i found:
- on the first picture, See the cursor left on the chart. Its on a building. In the original it shows the type of building size. On serfling it shows nothing. This is a porblem if you search for more castles on your map and wanna check the small buildings if you can use them for a big building.
- on the castle/warehouse page, ( last picture ) where i can choose, if men or material goes in, out or collect, in the original i need to press both mouse buttons to confirm any changes on that page except, if i turn it back to the first normal function.
- the mouse courser is missing the repetition when I keep the left mouse button held down. Its only working on the start screen. 

Nr.6:
[SAVE3.DS.ZIP](https://github.com/nicymike/serflings/files/5002393/SAVE3.DS.ZIP)
Its not a Zip, i just renamed it so it will stick in here. Just rename it again.
Then the warehouse to the north east of the main castle is stuck.

**Comment by nicymike** [comment]
Thank you for the answer. Now I know what you mean by knight school. Actually the number is the minimum number of knights defending the castle.

I moved some of the other issues to new entries because I need separate entries to plan issues for different game version. I will keep this entry for the knight training issue. Feel free to create additional entries for the remaining issues.

Now to my answers:

> Thanks for the detailed answer. I feel bad keeping you busy with that. :)

No worries

> So here for Nr.2 the screenshot; I think the problem appears above 100%, three digits.

Discussion moved to: https://github.com/nicymike/serflings/issues/22

> I ran a quick test again and the result, as you can see the big knights left on the original game and where still in school in serflings after the swap. So in the original they walk into the castle as freshmen and walk out as proper kinghts, while in serflings they walk in the same they walk out. There are still some strange things going on, which dont match in the school. :)

There are some differences to the original game in Serflings at different locations. That's why I need to validate all parts of the code. If you check the changelog you will see that some parts are already done. Knight exchange is one thing I still need to validate.

> * Another difference is, when I want to open school places, in serflings, knights are placed instantly into school while in the original it takes a bit a time to get em there. Same effect when you dismiss them from school.

Yes, this is already on my list. Also the worker for a new construction site leaves the castle without a delay. I think this is the same issue.

> * And above the diffrence when I recruit 20 of em back to school: Original puts the freshmen first. Serflings is taking the finished knights back into school.

Maybe this has something to do with knight exchange

> * on the first picture, See the cursor left on the chart. Its on a building. In the original it shows the type of building size available on that surface. On serfling it shows nothing. This is a porblem if you search for more castles on your map and wanna check the small buildings if you can use them for a big building.

Yes, this is because of the missing feature of replacing an existing building.

> * on the castle/warehouse page, ( last picture ) where i can choose, if men or material goes in, out or collect, in the original i need to press both mouse buttons to confirm any changes on that page except, if i turn it back to the first normal function.

Thank you for the clue. I didn't know this. You may create a separate issue for this.

> * the mouse courser is missing the repetition when I keep the left mouse button held down. Its only working on the start screen.

Do you mean repeated clicks? I did not add this because it can be very annoying. For what action do you need this?

> * sound: the metal melting guys have a crack sound in their loop.

I think this has something to do with the sound issue I mentioned earlier regarding the bird sounds.

> Nr.6:
> [SAVE3.DS.ZIP](https://github.com/nicymike/serflings/files/5002393/SAVE3.DS.ZIP)
> Its not a Zip, i just renamed it so it will stick in here. Just rename it again.
> Then the warehouse to the north east of the main castle is stuck.

Discussion moved to: https://github.com/nicymike/serflings/issues/23

> For the multiplayer mode, I cannot set the amount of goods for the players at start, always comes out with about 117 swords and shields. Thats really sad, i would love to have em on the minimum at least or at best individual changeable of course.But you learn the game best from the minimum goods available.

When I added multiplayer I did not want to add a complete interface because it is very time consuming. And because you cannot change the settings I set them to static values. If you want custom values you may create a new savegame in the original settlers 1 and then load it in Serflings. This even works for multiplayer games.
Also check out: https://github.com/nicymike/serflings/issues/12

**Comment by Wakeboardtiger** [comment]
> > * the mouse courser is missing the repetition when I keep the left mouse button held down. Its only working on the start screen.
> 
> Do you mean repeated clicks?  For what action do you need this?
Yes, exactly thats what i meant.
As mentioned on the start screen it is helpful to change the blue/green/red bars which works and in game:
-  to press +/- buttons for the 99 school or castle places available.
-  to change via +/-  how many knights will be going to attack a building.
-  on the page where you decide how many knights go into a building, its easy to put everything quick to full or minimum.
-  on the page where to decide which tools are made from the tool maker, to set all bars quickly to zero.
-  on the pages where to decide which good is being handled first, to move individual goods quickly up and down.
-  to click gathered millions of messages away quickly. 
All i can think of for now. :)


> > For the multiplayer mode, I cannot set the amount of goods for the players at start, always comes out with about 117 swords and shields. Thats really sad, i would love to have em on the minimum at least or at best individual changeable of course. But you learn the game best from the minimum goods available.
> 
> When I added multiplayer I did not want to add a complete interface because it is very time consuming. And because you cannot change the settings I set them to static values. If you want custom values you may create a new savegame in the original settlers 1 and then load it in Serflings. This even works for multiplayer games.
> Also check out: #12

Yo, I feared I was missing something. This does the trick, great to know, many thanks.

**Comment by nicymike** [comment]
About repeated clicks:
Well, you are right. I did completely forget the buttons where you can break your fingers by clicking hundreds of times. 😃

Can you please add a new issue for this? I think I can add this feature for individual buttons or the bars.

**Comment by nicymike** [comment]
Fix included in next development version:
The special click for stock management is needed if additional special click is enabled in config (specialClickAdditional).

**Comment by nicymike** [comment]
For the crack sound on smelter check out: https://github.com/nicymike/serflings/issues/5

**Comment by nicymike** [comment]
Training and knight exchange will be fixed in the next version. I will close this issue because the other topics have been moved to different issues.

Please create a new issue if I forgot something.
